sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -508,164 +498,202 @@ export class EventHubSender extends LinkEntity {
+                this.name,
+                delivery.id,
+                delivery.tag.toString()
+              );
             } else {
-              err = new Error(
+              // let us retry to send the message after some time.
+              const msg =
                 `[${this._context.connectionId}] Sender "${this.name}", ` +
-                  `received a modified disposition.Hence we are rejecting the promise.`
-              );
+                `cannot send the message right now. Please try later.`;
+              log.error(msg);
+              const amqpError: AmqpError = {
+                condition: ErrorNameConditionMapper.SenderBusyError,
+                description: msg
+              };
+              reject(translate(amqpError));
             }
-            log.error(err);
-            reject(err);
-          };
+          });
 
-          const actionAfterTimeout = () => {
-            removeListeners();
-            const desc: string =
-              `[${this._context.connectionId}] Sender "${this.name}" with ` +
-              `address "${this.address}", was not able to send the message right now, due ` +
-              `to operation timeout.`;
-            log.error(desc);
-            const e: AmqpError = {
-              condition: ErrorNameConditionMapper.ServiceUnavailableError,
-              description: desc
-            };
-            return reject(translate(e));
-          };
+        waitTimer = setTimeout(
+          actionAfterTimeout,
+          getRetryAttemptTimeoutInMs(options.retryOptions)
+        );
 
-          if (abortSignal) {
-            abortSignal.addEventListener("abort", onAborted);
-          }
-          this._sender!.on(SenderEvents.accepted, onAccepted);
-          this._sender!.on(SenderEvents.rejected, onRejected);
-          this._sender!.on(SenderEvents.modified, onModified);
-          this._sender!.on(SenderEvents.released, onReleased);
-          waitTimer = setTimeout(actionAfterTimeout, getRetryAttemptTimeoutInMs(options.retryOptions));
-          const delivery = this._sender!.send(message, undefined, 0x80013700);
+        if (!this.isOpen()) {
           log.sender(
-            "[%s] Sender '%s', sent message with delivery id: %d",
-            this._context.connectionId,
-            this.name,
-            delivery.id
+            "Acquiring lock %s for initializing the session, sender and " +
+              "possibly the connection.",
+            this.senderLock
           );
-        } else {
-          // let us retry to send the message after some time.
-          const msg =
-            `[${this._context.connectionId}] Sender "${this.name}", ` +
-            `cannot send the message right now. Please try later.`;
-          log.error(msg);
-          const amqpError: AmqpError = {
-            condition: ErrorNameConditionMapper.SenderBusyError,
-            description: msg
-          };
-          reject(translate(amqpError));
+          defaultLock
```

Comment:
It's fine to chain promises like this, but you can also use async/await if you make the function you pass into `new Promise()` an async function. It might make this section easier to reason about, but I don't have a strong opinion on it.